index.rst

```diff
@@ -42,20 +42,151 @@
+On completion they register themselves as inputs to a combination pipeline.
+This pipeline does not run automatically but waits to be told that all data have arrived.
+If an observation fails to reduce properly, or fails to pass some metrics threshold it will not register with the combination pipeline.
+
+When all expected data have arrived and been processed on their own, the final combination pipeline will be unblocked and will run to completion.
+Events will be issued on the completion of a single frame processing and on the completion of the combination pipeline.
+Metrics will also be published where appropriate during processing although the mechanism for this has not been defined.
+
+The issues that have to be solved are:
+
+1. How does the pipeline know to start processing a data ID?
+2. Which pipeline should be run for the given data ID?
+3. If the pipeline supports chaining, is that known to the system passing in the data?
+4. Who decides which observations should be combined and how?
+5. How does the system know the final observation arrived such that the combination pipeline can run?
+
+This document will describe two alternate scenarios for answering the above questions.
+The first describes a system entirely controlled by the observing script that is running the observation.
+The second describes a system that is decoupled from the observing system and is entirely driven by data arriving in the system.
+
+To simplify the remainder of this document the name given to the system that forwards processing requests to pipelines will be the "OCS Controlled Pipeline System (OCPS)."
+
+Observing Script Control
+========================
+
+In this scenario the observing script or notebook [*]_ has full control of all processing and the observing sequence and data processing are tightly coupled.
+The control flow could look something like this:
+
+* Tell OCPS that a new co-addition pipeline with the defined name is to be readied.
+* Loop over however many images are to be taken
+   * Send the "take images" command to the camera, including the group ID issued by the ScriptQueue.
+   * Asynchronously wait for the "data available" message to come from the Archiver or CCS containing the obsid.
+      - Send command to OCPS requesting that data with that obsid be processed.
+
+* Once loop completes and all obsids have been sent for processing, send command to OCPS telling it to run the co-addition pipeline.
+
+If the script is told to stop early the script can send the command to run the co-addition pipeline as part of its clean up phase.
+
+In this system the group ID is not used to determine whether an observation should be co-added and its role is solely to indicate that the data were taken by that particular execution of the script.
+The scripting approach does not store any additional metadata with the data and therefore subsequent offline processing can not easily reproduce exactly the grouping specified by the observing script.
```

Comment:
The co-addition pipeline should save that provenance for exact reproduction, shouldn't it?

Reply:
Yes, so you can't reproduce the processing by looking at the data you took, you can only work it out by looking at the data products.